sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs

```diff
@@ -0,0 +1,275 @@
+                case AmqpProperty.Type.Single:
+                case AmqpProperty.Type.Double:
+                case AmqpProperty.Type.Boolean:
+                case AmqpProperty.Type.Decimal:
+                case AmqpProperty.Type.Char:
+                case AmqpProperty.Type.Guid:
+                case AmqpProperty.Type.DateTime:
+                case AmqpProperty.Type.String:
+                    amqpPropertyValue = eventPropertyValue;
+                    break;
+
+                case AmqpProperty.Type.Stream:
+                case AmqpProperty.Type.Unknown when eventPropertyValue is Stream:
+                    amqpPropertyValue = ReadStreamToArraySegment((Stream)eventPropertyValue);
+                    break;
+
+                case AmqpProperty.Type.Uri:
+                    amqpPropertyValue = new DescribedType(AmqpProperty.UriName, ((Uri)eventPropertyValue).AbsoluteUri);
+                    break;
+
+                case AmqpProperty.Type.DateTimeOffset:
+                    amqpPropertyValue = new DescribedType(AmqpProperty.DateTimeOffsetName, ((DateTimeOffset)eventPropertyValue).UtcTicks);
+                    break;
+
+                case AmqpProperty.Type.TimeSpan:
+                    amqpPropertyValue = new DescribedType(AmqpProperty.TimeSpanName, ((TimeSpan)eventPropertyValue).Ticks);
+                    break;
+
+                case AmqpProperty.Type.Unknown:
+                    var exception = new SerializationException(String.Format(CultureInfo.CurrentCulture, Resources.FailedToSerializeUnsupportedType, eventPropertyValue.GetType().FullName));
+                    EventHubsEventSource.Log.UnexpectedException(exception.Message);
+                    throw exception;
+            }
+
+            return (amqpPropertyValue != null);
+        }
+
+        /// <summary>
+        ///   Gets the AMQP property type identifier for a given
+        ///   value.
+        /// </summary>
+        ///
+        /// <param name="value">The value to determine the type identifier for.</param>
+        ///
+        /// <returns>The <see cref="AmqpProperty.Type"/> that was identified for the given <paramref name="value"/>.</returns>
+        ///
+        private static AmqpProperty.Type GetTypeIdentifier(object value) =>
+            (value == null)
+                ? AmqpProperty.Type.Null
+                : value.GetType().ToAmqpPropertyType();
+
+        /// <summary>
+        ///   Converts a stream to a set of memory bytes.
+        /// </summary>
+        ///
+        /// <param name="stream">The stream to read and capture in memory.</param>
+        ///
+        /// <returns>The set of memory bytes containing the stream data.</returns>
+        ///
+        private static ArraySegment<byte> ReadStreamToArraySegment(Stream stream)
+        {
+            if (stream == null)
+            {
+                return new ArraySegment<byte>();
+            }
+
+            using (var memStream = new MemoryStream(512))
```

Comment:
512 was the value used by the track one implementation.   Not completely sure why it was chosen or if it was a good idea to bring it forward.